pkg/controller/workspacetemplate/workspacetemplate_controller.go

```diff
@@ -18,387 +18,298 @@ package workspacetemplate
-			Name:        workspaceTemplate.Name,
-			Labels:      workspaceTemplate.Spec.Template.Labels,
-			Annotations: workspaceTemplate.Spec.Template.Annotations,
-		},
-		Spec: workspaceTemplate.Spec.Template.Spec,
-	}
-
-	err := controllerutil.SetControllerReference(workspaceTemplate, workspace, scheme.Scheme)
-	if err != nil {
+func (r *WorkspaceTemplateReconciler) createWorkspace(ctx context.Context, logger logr.Logger, workspace *tenantv1alpha1.Workspace) error {
+	if err := r.Create(ctx, workspace); err != nil {
+		logger.Error(err, "create workspace failed")
 		return err
 	}
+	return nil
+}
 
-	_, err = c.ksClient.TenantV1alpha1().Workspaces().Create(workspace)
-	if err != nil {
-		if errors.IsAlreadyExists(err) {
-			return nil
-		}
-		klog.Error(err)
+func (r *WorkspaceTemplateReconciler) updateWorkspace(ctx context.Context, logger logr.Logger, workspace *tenantv1alpha1.Workspace) error {
+	if err := r.Update(ctx, workspace); err != nil {
+		logger.Error(err, "update workspace failed")
 		return err
 	}
-
 	return nil
 }
 
-func (c *controller) updateWorkspace(workspace *tenantv1alpha1.Workspace) error {
-	_, err := c.ksClient.TenantV1alpha1().Workspaces().Update(workspace)
-	if err != nil {
-		klog.Error(err)
+func (r *WorkspaceTemplateReconciler) updateFederatedWorkspace(ctx context.Context, logger logr.Logger, federatedWorkspace *typesv1beta1.FederatedWorkspace) error {
+	unstructured := &unstructured.Unstructured{}
+	if obj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(federatedWorkspace); err != nil {
+		logger.Error(err, "convert to unstructured object failed")
+		return err
+	} else {
+		unstructured.Object = obj
+	}
+	logger.V(4).Info("update federated workspace")
+	if err := r.Update(ctx, unstructured); err != nil {
+		logger.Error(err, "update federated workspace failed")
 		return err
 	}
 	return nil
 }
 
-func (c *controller) initRoles(workspace *tenantv1alpha2.WorkspaceTemplate) error {
-	roleBases, err := c.roleBaseLister.List(labels.Everything())
-	if err != nil {
-		klog.Error(err)
+func (r *WorkspaceTemplateReconciler) initWorkspaceRoles(ctx context.Context, logger logr.Logger, workspace *tenantv1alpha2.WorkspaceTemplate) error {
+	var templates iamv1alpha2.RoleBaseList
+	if err := r.List(ctx, &templates); err != nil {
+		logger.Error(err, "list role base failed")
 		return err
 	}
-	for _, roleBase := range roleBases {
-		var role iamv1alpha2.WorkspaceRole
-		if err = yaml.NewYAMLOrJSONDecoder(bytes.NewBuffer(roleBase.Role.Raw), 1024).Decode(&role); err == nil && role.Kind == iamv1alpha2.ResourceKindWorkspaceRole {
-			roleName := fmt.Sprintf("%s-%s", workspace.Name, role.Name)
-			if role.Labels == nil {
-				role.Labels = make(map[string]string, 0)
+	for _, template := range templates.Items {
+		var expected iamv1alpha2.WorkspaceRole
+		if err := yaml.NewYAMLOrJSONDecoder(bytes.NewBuffer(template.Role.Raw), 1024).Decode(&expected); err == nil && expected.Kind == iamv1alpha2.ResourceKindWorkspaceRole {
```

Comment:
The `err` may need to be logged properly

Reply:
done